src/interpreter/compiler.ts

```diff
@@ -0,0 +1,290 @@
+  UnsupportedOperatorError,
+  UnsupportedOperatorErrorType
+} from './errors';
+import {
+  constructConditionalExpression,
+  constructFalseConstant,
+  constructMainCallExpression,
+  constructTrueConstant
+} from '../ast/constructors';
+import { type SymbolTable } from './types/symbolTable';
+import {
+  addBlockSymbolTableEntries,
+  addFunctionSymbolTableEntries,
+  addProgramSymbolTableEntries,
+  getFunctionSymbolTableEntry,
+  getNumOfEntriesInFrame,
+  getSymbolTableEntry,
+  getSymbolTableEntryInFrame
+} from './symbolTable';
+import { type Instr } from './types/instruction';
+
+export const compileProgram = (ast: Program): Instr[] => {
+  const symbolTable: SymbolTable = {
+    head: {},
+    tail: null,
+    parent: null
+  };
+  const state: CompilerState = {
+    symbolTable,
+    instructions: []
+  };
+  compile(ast, state);
+  const mainCallExpression = constructMainCallExpression();
+  compile(mainCallExpression, state);
+  const doneInstr = constructDoneInstr();
+  state.instructions.push(doneInstr);
+  return state.instructions;
+};
+
+const compile = (node: Node, state: CompilerState): void => {
+  // The typecast allows for mapping to a specific evaluator instr type from their union type.
+  // https://stackoverflow.com/questions/64527150/in-typescript-how-to-select-a-type-from-a-union-using-a-literal-type-property
+  compilers[node.type](node as any, state);
+};
+
+const compilers: CompilerMapping = {
+  ArrayAccessExpression: (
+    node: ArrayAccessExpression,
+    state: CompilerState
+  ) => {},
+  AssignmentExpression: (
+    node: AssignmentExpression,
+    state: CompilerState
+  ) => {},
+  BinaryExpression: (node: BinaryExpression, state: CompilerState) => {
+    compile(node.left, state);
+    compile(node.right, state);
+    const binaryOperationInstr = constructBinaryOperationInstr(node.operator);
+    state.instructions.push(binaryOperationInstr);
+  },
+  BlockStatement: (node: BlockStatement, state: CompilerState) => {
+    const blockSymbolTable = addBlockSymbolTableEntries(
+      node,
+      state.symbolTable
+    );
+    node.items.forEach((item) => {
+      state.symbolTable = blockSymbolTable;
+      compile(item, state);
+    });
+  },
+  BreakStatement: (node: BreakStatement, state: CompilerState) => {},
+  CallExpression: (node: CallExpression, state: CompilerState) => {
+    if (!isIdentifier(node.callee)) {
+      throw new InvalidCallError('Cannot call non-identifier.');
+    }
+    const functionEntry = getFunctionSymbolTableEntry(
+      node.callee.name,
+      state.symbolTable
+    );
+    if (functionEntry.numOfParams !== node.arguments.length) {
+      throw new InvalidCallError(
+        `Function takes in ${functionEntry.numOfParams} arguments but ${node.arguments.length} arguments were passed in.`
+      );
+    }
+
+    compile(node.callee, state);
+    // Compile in reverse order so that last argument is lower in the stash,
+    // and the first argument is higher.
```

Comment:
Can we specify that the stash here refers to the virtual machine's stash?

Also, this means we evaluate from right to left, right? No action needed for this, just want to be clear so that we can reference this PR in the future.